ffi/custom_passes.cpp

```diff
@@ -0,0 +1,613 @@
+        if (!md)
+            return false;
+        if (md->getNumOperands() != 1)
+            return false;
+        auto &operand = md->getOperand(0);
+        auto data = dyn_cast<ConstantAsMetadata>(operand.get());
+        if (!data)
+            return false;
+        return data->getValue()->isOneValue();
+    }
+
+    template<class T>
+    bool basicBlockInList(const BasicBlock* bb, const T &list){
+        for (BasicBlock *each : list) {
+            if (bb == each) return true;
+        }
+        return false;
+    }
+
+    bool hasDecrefInNode(CallInst* incref, BasicBlock* bb){
+        for (Instruction &ii : *bb) {
+            if (isRelatedDecref(incref, &ii) != NULL) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    template <class T>
+    bool eraseNullFirstArgFromList(T& refops) {
+        bool mutated = false;
+        for (CallInst*& refop: refops) {
+            if (!isNonNullFirstArg(refop)) {
+                refop->eraseFromParent();
+                mutated |= true;
+                refop = NULL;
+            }
+        }
+        return mutated;
+    }
+
+    /**
+     * Find related decrefs to incref inside a basicblock in order
+     */
+    std::vector<CallInst*> findRelatedDecrefs(BasicBlock* bb, CallInst* incref) {
+        std::vector<CallInst*> res;
+        for (Instruction &ii : *bb) {
+        CallInst *call_inst;
+        if ((call_inst = isRelatedDecref(incref, &ii))){
+            res.push_back(call_inst);
+        } else {
+            continue;
+        }
+        }
+        return res;
+    }
+
+    CallInst* isRelatedDecref(CallInst *incref, Instruction *ii) {
+        // TODO: DRY
+        if (ii->getOpcode() == Instruction::Call) {
+            CallInst *call_inst = dyn_cast<CallInst>(ii);
+            Value *callee = call_inst->getCalledOperand();
+            if ( callee->getName() != "NRT_decref" ) {
+                return NULL;
+            }
+            if (incref->getArgOperand(0) != call_inst->getArgOperand(0)) {
+                return NULL;
+            }
+            return call_inst;
+        }
+        return NULL;
+    }
+
+    void getAnalysisUsage(AnalysisUsage &Info) const override {
+        Info.addRequired<DominatorTreeWrapperPass>();
+        Info.addRequired<PostDominatorTreeWrapperPass>();
+    }
+
+    bool isNonNullFirstArg(CallInst *call_inst){
+        auto val = call_inst->getArgOperand(0);
+        auto ptr = dyn_cast<ConstantPointerNull>(val);
+        return ptr == NULL;
+    }
+
+    bool hasAnyDecrefInNode(BasicBlock *bb) {
+
+        for (Instruction &ii: *bb) {
+            CallInst* refop = GetRefOpCall(&ii);
+            if (refop != NULL && IsDecRef(refop)) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Pre-condition: head_node dominates tail_node
+     */
+    bool hasDecrefBetweenGraph(BasicBlock *head_node, BasicBlock *tail_node,
```

Comment:
i'm still in the mode of _just getting it to work_. There are probably a few inefficient code patterns.

Reply:
From OOB chat, this needs to be made iterative or have a recursion limit else it could overflow the stack for a pathological IR input e.g `<incref><massive if/else tree><decref>`